File: JIM_setup/controller_JIM_test_TCP.m
data = readtable('ankle_test_right_swing_112run1.csv', 'VariableNamingRule','modify');
freq = 250;
% sample_num = height(data);
sample_num = 1000;
period = 1/freq;

% TCP Port
port = 12345;
ip = '35.3.141.14';
ip_client = tcpclient(ip, port,"Timeout", 0.1);
disp('Client is connected to server.');

% Load Data
data_str = cell(sample_num, 1);
for i = 1:sample_num
    data_str{i} = sprintf('%g,%g', data.AnkleAngle(i), data.ControllerTorque(i));
end
disp('Data Loaded');

% Send Data
tic;    % Start Timing
i = 1;
next_time = 0;
while i <= sample_num
    write(ip_client, data_str{i});

    next_time = next_time + period;
    elapsed_time = toc;
    while elapsed_time < next_time
        pause(1e-6);
        elapsed_time = toc;
    end
    i = i + 1;
end

writeline(ip_client, "END");

elapsed_time = toc; % End Timing
send_freq = sample_num/elapsed_time;
fprintf('Total transmission time: %.3f seconds\n', elapsed_time);
fprintf('Transmission frequency: %.3f Hz\n', send_freq);

clear ip_client;
disp('Client disconnected from server.');

File: JIM_setup/controller_JIM_test_TCP.py
import socket
import time
import pandas as pd

data = pd.read_csv('I:\\My Drive\\Locomotor\\ExoBoot\\JIM_setup\\ankle_test_right_swing_112run1.csv')

freq = 250  # Hz
sample_num = 1000
period = 1 / freq

remote_port = 12345
remote_ip = '35.3.141.14'
end_marker = 'END'

s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
s.connect((remote_ip, remote_port))
print('TCP client is connected to the server.')

data_str = []
for i in range(sample_num):
    data_str.append(f"{data['Ankle Angle'][i]},{data['Controller Torque'][i]}")


start_time = time.time()
for i in range(sample_num):
    s.send(data_str[i].encode())
    elapsed_time = time.time() - start_time
    sleep_time = period - elapsed_time % period
    if sleep_time > 0:
        time.sleep(sleep_time)
s.send(end_marker.encode())


elapsed_time = time.time() - start_time
send_freq = sample_num / elapsed_time
print(f'Total transmission time: {elapsed_time:.3f} seconds')
print(f'Transmission frequency: {send_freq:.3f} Hz')

s.close()
print('TCP sender disconnected from the server.')